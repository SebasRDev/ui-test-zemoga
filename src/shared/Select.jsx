import PropTypes from "prop-types";
import "../styles/shared/Select.css";
import { useState } from "react";

export const Select = ({options, selected, onChange}) => {

  const [isOpen, setIsOpen] = useState(false);

  const handleChange = (value) => {
    onChange(value);
    setIsOpen(false);
  }

  return (
    <div className="select">
      <div className="select__item" onClick={() => setIsOpen(!isOpen)}>
        <p>{selected}</p>
        <svg
          width="11"
          height="7"
          viewBox="0 0 11 7"
          fill="none"
        >
          <path
            fillRule="evenodd"
            clipRule="evenodd"
            d="M5.25 7L6.11959e-07 -9.17939e-07L10.5 0L5.25 7Z"
            fill="#333333"
          />
        </svg>
      </div>
      <div className={`select__options ${isOpen &&'select__options--open'}`}>
        {options.map(({value, label}) => {
          return (
            <div className="select__item" key={value} onClick={() => handleChange(value)}>
              {label}
            </div>
          );
        })}
      </div>
    </div>
  );
};

Select.propTypes = {
  selected: PropTypes.string.isRequired,
  options: PropTypes.array.isRequired,
  onChange: PropTypes.func.isRequired,
};
