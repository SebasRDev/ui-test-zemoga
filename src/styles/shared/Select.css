.select{
  width: 130px;
  height: 36px;
  position: relative;
  z-index: 100;
}

.select__item{
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 8px 16px 8px 8px;
  border: 2px solid #333;
  cursor: pointer;

  p{
    display: block;
    width: 90px;
    text-align: center;
    margin: 0;
    text-transform: capitalize;
  }
}

.select__options{
    position: absolute;
    left: 0;
    width: 100%;
    background-color: #fff;
    display: none;
}

.select__options--open{
    display: block;
}

.select__options > div {
    border-top: none;
    justify-content: center;

    &:hover{
      background-color: #f2f2f2;
    }
}