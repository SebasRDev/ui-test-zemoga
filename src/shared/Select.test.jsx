import { describe, expect, it, beforeEach, vi } from "vitest";
import { render, screen, fireEvent, act, waitFor } from "@testing-library/react";
import { Select } from "./Select";

const mockLayoutOptions = [
  { value: "list", label: "List" },
  { value: "grid", label: "Grid" },
];

// Simular localStorage
window.localStorage = {
  getItem: vi.fn(),
  setItem: vi.fn(),
};

describe("Select", () => {
  beforeEach(() => {
    render(<Select selected="list" options={mockLayoutOptions} onChange={() => {}}/>);
  });

  it("should render a select component", () => {
    const { container } = render(<Select selected="list" options={mockLayoutOptions} onChange={() => {}}/>);
    expect(container).toMatchSnapshot();
  });

  it("should show select options", () => {
    const select = screen.getByTestId("test-select");
    fireEvent.click(select);
    const selectOptions = screen.getByTestId("test-select-options");
    expect(selectOptions).toHaveClass("select__options--open");
  });

  it("should change the layout", () => {
    const select = screen.getByTestId("test-select");
    fireEvent.click(select);
    const selectOptions = screen.getAllByRole("select-options");
    fireEvent.click(selectOptions[1]);
    expect(selectOptions[1]).toHaveTextContent("Grid");
  });

  it("should change the layout in localstorage", async() => {
    const select = screen.getByTestId("test-select");
    fireEvent.click(select);
    const selectOptions = screen.getAllByRole("select-options");
    act(() => {
      fireEvent.click(selectOptions[1]);
    })
    waitFor(() => {
      expect(localStorage.setItem).toHaveBeenCalledWith("layout", "grid");
    });
  });
});
