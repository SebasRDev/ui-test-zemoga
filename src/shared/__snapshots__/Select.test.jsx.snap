// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select > should render a select component 1`] = `
<div>
  <div
    class="select"
  >
    <div
      class="select__item"
      data-testid="test-select"
    >
      <p>
        list
      </p>
      <svg
        fill="none"
        height="7"
        viewBox="0 0 11 7"
        width="11"
      >
        <path
          clip-rule="evenodd"
          d="M5.25 7L6.11959e-07 -9.17939e-07L10.5 0L5.25 7Z"
          fill="#333333"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <div
      class="select__options false"
      data-testid="test-select-options"
    >
      <div
        class="select__item"
        role="select-options"
      >
        List
      </div>
      <div
        class="select__item"
        role="select-options"
      >
        Grid
      </div>
    </div>
  </div>
</div>
`;
